services/vpm/vpm-version.service.js

```diff
@@ -0,0 +1,81 @@
+import Joi from 'joi'
+import { optionalUrl } from '../validators.js'
+import { latest, renderVersionBadge } from '../version.js'
+import { BaseJsonService, NotFound } from '../index.js'
+
+const queryParamSchema = Joi.object({
+  repositoryUrl: optionalUrl.required(),
+  sort: Joi.string().valid('defined', 'semver').default('defined'),
+}).required()
+
+const schema = Joi.object({
+  packages: Joi.object()
+    .pattern(
+      /./,
+      Joi.object({
+        versions: Joi.object().pattern(/./, Joi.object()).required(),
+      }).required()
+    )
+    .required(),
+}).required()
+
+export default class VpmVersion extends BaseJsonService {
+  static category = 'version'
+
+  static route = {
+    base: 'vpm/v',
+    pattern: ':packageId',
+    queryParamSchema,
+  }
+
+  static examples = [
+    {
+      title: 'VPM Package Version',
+      namedParams: {
+        packageId: 'com.vrchat.udonsharp',
+      },
+      queryParams: {
+        repositoryUrl: 'https://packages.vrchat.com/curated?download',
+      },
+      staticPreview: renderVersionBadge({ version: '1.1.6' }),
+    },
+    {
+      title: 'VPM Package Version (latest semver)',
+      namedParams: {
+        packageId: 'com.vrchat.udonsharp',
+      },
+      queryParams: {
+        repositoryUrl: 'https://packages.vrchat.com/curated?download',
+        sort: 'semver',
+      },
+      staticPreview: renderVersionBadge({ version: '1.1.6' }),
+    },
+  ]
+
+  static defaultBadgeData = {
+    label: 'vpm',
+  }
+
+  async fetch({ repositoryUrl }) {
+    return this._requestJson({
+      schema,
+      url: repositoryUrl,
+    })
+  }
+
+  async handle({ packageId }, { repositoryUrl, sort }) {
+    const data = await this.fetch({ repositoryUrl })
+    const pkg = data.packages[packageId]
+    if (pkg === undefined)
+      throw new NotFound({ prettyMessage: 'package not found' })
+    const versions = Object.keys(pkg.versions)
+    let version
+    if (sort === 'defined') {
+      version = versions.reverse()[0]
+    } else if (sort === 'semver') {
+      version = latest(versions)
+    }
```

Comment:
Can you expand a bit on the reason for allowing multiple sort methods here? I don't really know the norms of this community, but for package registries we generally want the concept of 'latest' to align with that of the upstream registry (What version would I get if I run `vpm install`, or whatever?). We do have this for some badges like github tags/releases where the platform is unaware of the language ecosystem norms or version schema (a github release could be anything - an NPM package, an android app, whatever..). For package registries (NPM, Ruby gems, Rust crates, PyPI, etc) we don't generally do this.

Reply:
Also, normally for version badges, `/v` would be the latest version (excluding pre-releases) and we'd have an `include_prereleases` flag for latest version (including pre-releases). Does VPM have any concept of stable/pre-releases?

Reply:
Because official repository template is made based on GitHub releases, in most case, its ordered by release date. however, it's not documented & not a rule for the repository so we can remove this sort.

I think this sort of useful in case of releasing multiple major releases but it may not be a enough reason.

> Does VPM have any concept of stable/pre-releases?
Yes. VPM is based on semver. I forget to check prereleases in `defined` sort.